### ISSUE TITLE
Bot_id is different for message and user_list responses.

### ISSUE THREAD

**Issue by BootGenius** [slackapi/java-slack-sdk#846]
When I retrieve a list of users through the API I get a list of users and bots. The bot I am interested in has a bot_id=B02ETC8TFMX. But when retrieve a list of messages from the API and locate the messages made by that bot I have another bot_id (B02E81EKAUE) set for the message. How is that possible?
Messages are truly belong to this bot (I can understand it by unique message text).
Is it a bug in slack SDK or in the Slack platform in general?

**Comment by seratch** [comment]
Hi @BootGenius, thanks for writing in!

>Is it a bug in slack SDK or in the Slack platform in general?

This is a server-side behavior. **The `bot_id` in messages posted by slash command invocation can have a different bot_id value even for the same Slack app.** This situation can arise only with the ones posted by `ctx.ack("This message")` in Bolt slash command listeners. Under the hood, the `ack` method sends the response data as the HTTP response body in response to a slash command incoming request.

Our server-side is not planning to change the bot_id difference for slash commands. If your app development already relies on the assumption that only one `bot_id` can exist for a Slack app, I do understand that this is frustrating. However, we generally recommend referring to a bot's `user_id`, not `bot_id` to identify an app's bot user in message data.

I hope this helps!

**Comment by BootGenius** [comment]
Thanks for the answer. Your suggestion is right and we used it until we realized that for messages from the bots user_id is set to null in the message

**Comment by BootGenius** [comment]
I mean in message event or when I retrieve a list of messages the user_id of bot's message is set to null always

**Comment by seratch** [comment]
Indeed, due to the absence of user_id, there is no way to know which app sent the message.

I don't know how much this is useful for you, but, by calling bots.info API method with the bot_id, you can get the app's app_id. If your database manages the list of installed apps' IDs, it would be possible to identify the message poster. I know this is not straight-forward at all but wanted to share something I can think of right now.

I'll escalate this feedback (=the absence of user_id in message data) to the server-side engineering teams. 

We are sorry about the inconvenience here but we are unable to manage the progress of the server-side tasks here. For the inquires of the server-side updates, please contact our customer support agents.

**Comment by BootGenius** [comment]
Ok. Got you. Thanks for the answer

**Comment by seratch** [comment]
@BootGenius I've talked to the server-side teams and now they recognize the issue. Unfortunately, making the `bot_id` consistent is not possible for the reasons such as backward compatibility for Slack desktop / mobile client apps. Therefore, I proposed adding `user_id` to the message data in this scenario. The task to add `user_id` is filed in our internal issue tracker. I cannot tell when the improvement will be available for you but I will continue communications for its fast delivery.

As I mentioned above, we are unable to share the server-side progress here in a timely manner. Let me close our discussion on the SDK project side. Thanks again for your feedback!
